src/main/java/com/iexec/core/task/update/TaskUpdateRequestManager.java

```diff
@@ -44,15 +45,30 @@
     private static final int TASK_UPDATE_THREADS_POOL_SIZE = Runtime.getRuntime().availableProcessors() * 2;
 
     private final ExecutorService executorService = Executors.newFixedThreadPool(1);
-    private final BlockingQueue<String> queue = new LinkedBlockingQueue<>();
     private final ConcurrentMap<String, Object> locks = ExpiringMap.builder()
             .expiration(LONGEST_TASK_TIMEOUT, TimeUnit.HOURS)
             .build();
-    private final Executor taskUpdateExecutor = TaskExecutorUtils.newThreadPoolTaskExecutor(
-            "task-update-",
-            TASK_UPDATE_THREADS_POOL_SIZE);
+
+    final TaskUpdatePriorityBlockingQueue queue = new TaskUpdatePriorityBlockingQueue();
+    // Both `corePoolSize` and `maximumPoolSize` should be set to `TASK_UPDATE_THREADS_POOL_SIZE`.
+    // Otherwise, `taskUpdateExecutor` won't pop `maximumPoolSize` threads
+    // as new threads are popped only if the queue is full
+    // - which never happens with an unbounded queue.
+    final ThreadPoolExecutor taskUpdateExecutor = new ThreadPoolExecutor(
+            TASK_UPDATE_THREADS_POOL_SIZE,
+            TASK_UPDATE_THREADS_POOL_SIZE,
+            0,
+            TimeUnit.MILLISECONDS,
+            queue
+    );
     private TaskUpdateRequestConsumer consumer;
```

Comment:
I don't really like the kind of circular dependency between `TaskUpdateRequestManager` and `TaskUpdateManager` which implements `TaskUpdateRequestConsumer` and is the bean stored in `consumer` I see.
They seem to be fundamentally intertwined.

Reply:
Sure, I find it strange, too. I wonder whether `TaskUpdateManager#publishUpdateTaskRequest` should be into `TaskUpdateRequestManager`. BTW, why not using the already existing `TaskUpdateRequestManager#publishRequest` method?
@jeremyjams or @zguesmi, do you have any clue on why it has been designed this way?

Reply:
The initial idea was to expose a single `TaskUpdateManager` entrypoint to others services (`TaskUpdateManager#publishUpdateTaskRequest`). Internal updates would be handle by a dedicated class (`TaskUpdateRequestManager`) in order to avoid overloading `TaskUpdateManager` more and make testing easier.

What do you propose?

Reply:
I see the point.
As of now, there's only a few public methods in `TaskUpdateManager`:
- `onTaskUpdateRequest`, called from `TaskUpdateRequestManager`,
- `publishUpdateTaskRequest`, calling `TaskUpdateRequesterManager#publishRequest`;
- `isConsensusReached`, called from `ReplicateSupplyService` and `TaskUpdateManager` itself.

Regarding these 3 methods, I don't think `TaskUpdateManager` is an entrypoint anymore.
What about the following:
- `TaskUpdateRequestManager` becomes this entrypoint and directly calls `TaskUpdateManager#updateTask`;
- methods that currently call `TaskUpdateManager#publishUpdateTaskRequest` call `TaskUpdateRequestManager#publishRequest` instead;
- `isConsensusReached` is moved to `TaskService`: almost all classes depending on `TaskUpdateManager` also depend on `TaskService` so that does not look to be a big deal;
- `onTaskUpdateRequest` and `publishUpdateTaskRequest` methods are deleted, and so is the interface `TaskUpdateRequestConsumer`.

Reply:
Yes, it's fine to me. Just make sure previous points are addressed before refactoring

Reply:
I'd like to work on it once this PR is merged. That would probably be cleaner :)

Reply:
Seems a good idea to me, both on the architecture side and on implementing this in another PR

Reply:
+1 to what you're suggesting @mcornaton .

Reply:
Suggestion implemented in https://github.com/iExecBlockchainComputing/iexec-core/pull/484.